### DISCUSSION TITLE
Typescript + Styledcomponent  + Navlink in charkra ?

### DISCUSSION THREAD

**Discussion by Emiltayeb** [chakra-ui/chakra-ui#4599] [Q&A]
Hello everyone.

My purpose is to create a custom navigation Navlink using a styled component and to utilize the *fontSize* attribute from chakra so that I don't need to modify CSS.
I want Navlink  becuase of the `activeclassname` react handel for me.

**The main golal is like this:**

```jsx

      <StyledLink
        borderColor={linkBorderColor}
        to={to}
        fontSize={['0.5rem', '3rem']}>
  
       
          {/* some text.. */}
    
      </StyledLink>
```

So what I did was passed Navlink form the `react router dom` to the Styled component and simply made the following calls:

```typescript

import { NavLink, NavLinkProps } from 'react-router-dom';

interface StyledLinkProps {
  borderColor: string;
}

const StyledLink = styled(NavLink)<NavLinkProps | StyledLinkProps>`
  border-bottom: 1px transparent solid;
  padding-bottom: 0.5rem;
  margin: 0 1rem;

  transition: border-bottom ease 0.5s;
  &.active,
  &:hover {
    border-color: ${({ borderColor }) => borderColor};
  }
`;
```
but i got this error:

![image](https://user-images.githubusercontent.com/49617556/130236069-c3e7707d-125c-4b52-85c8-e8cb7be6e455.png)

`   Property 'fontSize' does not exist on type 'IntrinsicAttributes & (((Omit<Omit<NavLinkProps<unknown> & RefAttributes<HTMLAnchorElement> & StyledLinkProps, never> & Partial<...>, "theme"> & { ...; }) | (Omit<...> & { ...; })) & { ...; })'.`

I understand the `Navlink` dosent know about the `fontSize` attribute. but how do i make the `Navlink` extends the chakra features?.

what i manneged to do which is working is the folowing:

```typescript


const StyledLink = styled(({ borderColor, ...rest }) => <NavLink {...rest} />)`
  border-bottom: 1px transparent solid;
  padding-bottom: 0.5rem;
  margin: 0 1rem;

  transition: border-bottom ease 0.5s;
  &.active,
  &:hover {
    border-color: ${({ borderColor }) => borderColor};
  }
`;
```

But now, i lose the type control over the `borderColor` prop.

Is there any way to make this work?
Thank you very much.

**Comment by chasinhues** [comment]
You don't need to use styled components, it can all be done with Chakra! :D 

https://codesandbox.io/s/pensive-maxwell-drqx2?file=/src/theme/index.ts

☝️ I threw together this demo for you. It shows how you can achieve what you want by extending the theme to add a new variant for Chakra's `Link` component. That covers the styling for you in a centralised place that you can reuse anywhere!

As for getting the appropriate types from both `NavLink` and Chakra's `Link`, you can use the `as` prop for that! All in the demo :)

If you have any extra questions about what I've done in the demo, let me know, happy to help!

**Reply by Emiltayeb** [reply]
Thank you very much for this detailed answer.

i'm looking at the code right now, I'm new to chakra and haven't really used the `extendTheme` yet, but it seems pretty straightforward.

i dont undersrand why im not getting auto completion for my styled. thats weird..
![image](https://user-images.githubusercontent.com/49617556/130316037-667d8f47-753e-4113-bb44-e8bb5131c586.png)


i saw that in `index.ts` inside the theme folder you did:

```typescript
const theme = extendTheme({
  components: {
    Link: {
      variants: {
        'styled': ({colorScheme}: Record<string, any>) => ({
         
        })
      }
    }
  }
})
```

Can you kindly explain why we need the `Record` here? i know its more Typescript realted but im curious.

**Reply by chasinhues** [reply]
Ah yeah you can strengthen the typings there for sure. 

I updated the sandbox link: https://codesandbox.io/s/pensive-maxwell-drqx2?file=/src/theme/components/Link.ts

Move the `Link` theme extension into it's own file, and type it with the `ComponentSingleStyleConfig` type. That should give you auto completion. 

New `/theme/components/Link.ts`:
```jsx
import { ComponentSingleStyleConfig } from "@chakra-ui/theme";

const Link: ComponentSingleStyleConfig = {
  variants: {
    styled: ({ colorScheme }) => ({
      borderBottom: "1px solid transparent",
      paddingBottom: "2",
      marginY: "0",
      marginX: "4",
      transition: "all ease 0.5s",
      _hover: {
        textDecoration: "none",
        borderColor: `${colorScheme}.500`
      },
      _active: {
        borderColor: `${colorScheme}.500`
      }
    })
  }
};

export default Link;

```

Updated `/theme/index.ts`:
```jsx
import { extendTheme } from "@chakra-ui/react";
import Link from "./components/Link";

const theme = extendTheme({
  components: {
    Link
  }
});

export default theme;

```
